### PR TITLE
Split locals and args

### PR DESCRIPTION
**Pull request by nbhuiyan**
Signed-off-by: Nazim Uddin Bhuiyan <nazimudd@ualberta.ca>

### CODE REVIEW

File: b9/src/ExecutionContext.cpp [hunk 1]
```diff
@@ -223,7 +228,15 @@ void ExecutionContext::doPushFromVar(StackElement *args, Immediate offset) {
   stack_.push(args[offset]);
 }
 
-void ExecutionContext::doPushIntoVar(StackElement *args, Immediate offset) {
+void ExecutionContext::doPopIntoVar(StackElement *args, Immediate offset) {
+  args[offset] = stack_.pop();
+}
+
+void ExecutionContext::doPushFromArg(StackElement *args, Immediate offset){
```
Comment (rwy7) [inline b9/src/ExecutionContext.cpp]: formatting

File: b9/src/ExecutionContext.cpp [hunk 2]
```diff
@@ -223,7 +228,15 @@ void ExecutionContext::doPushFromVar(StackElement *args, Immediate offset) {
...
+  stack_.push(args[offset]);
+}
+
+void ExecutionContext::doPopIntoArg(StackElement *args, Immediate offset){
```
Comment (rwy7) [inline b9/src/ExecutionContext.cpp]: formatting

File: b9/src/ExecutionContext.cpp [hunk 3]
```diff
@@ -119,8 +119,7 @@ StackElement ExecutionContext::interpret(const std::size_t functionIndex) {
         doPushFromVar(args, instructionPointer->immediate());
         break;
       case OpCode::POP_INTO_VAR:
-        // TODO bad name, push or pop?
-        doPushIntoVar(args, instructionPointer->immediate());
+        doPopIntoVar(args, instructionPointer->immediate());
```
Comment (rwy7) [inline b9/src/ExecutionContext.cpp]: *sigh* what a silly mistake 🐱
thanks for fixing.

File: b9/include/b9/instructions.hpp [hunk 4]
```diff
@@ -70,6 +70,12 @@ enum class OpCode : RawOpCode {
   // Jump if two strings are not equal
   STR_JMP_NEQ = 0x17,
 
+  //Push from function parameter
+  PUSH_FROM_ARG = 0x18,
+
+  //Pop into a parameter
+  POP_INTO_ARG = 0x19,
+
```
Comment (rwy7) [inline b9/include/b9/instructions.hpp]: I know this is a lot of tedious work, but can we have the push/pop from arg come right after (or before) the var instructions?

File: b9/src/MethodBuilder.cpp [hunk 5]
```diff
@@ -335,11 +335,13 @@ bool MethodBuilder::generateILForBytecode(
   }
 
   switch (instruction.opCode()) {
+    case OpCode::PUSH_FROM_ARG:
     case OpCode::PUSH_FROM_VAR:
       push(builder, loadVarIndex(builder, instruction.immediate()));
       if (nextBytecodeBuilder)
         builder->AddFallThroughBuilder(nextBytecodeBuilder);
       break;
+    case OpCode::POP_INTO_ARG:
```
Comment (rwy7) [inline b9/src/MethodBuilder.cpp]: OK, looking at the frontend, the parameters and locals each use a different symbol table. This means that args and params are indexed differently; arg[0] is not the same as local[0]. But here, it looks like arg[0] and local[0] are aliased. Is this a bug? There's the same issue in the interpreter.

File: b9run/main.cpp [hunk 6]
```diff
@@ -35,7 +35,7 @@ static const char* usage =
 struct RunConfig {
   b9::Config b9;
   const char* moduleName = "";
-  const char* mainFunction = "<body>";
+  const char* mainFunction = "b9main";
```
Comment (rwy7) [inline b9run/main.cpp]: why did we change this?

**Review by rwy7 — CHANGES_REQUESTED** [review]
IIUC, there are some issues with arguments and parameters being overlayed on the same stack slots.

On top of that, I would really like to get our terminology consistent, and I think this PR is a good place to start:

1) parameters vs arguments: I suggest going with param or parameter everywhere. It's closer to the JS standard and esprima terminology. parameter used to be an slightly overloaded term, but instruction parameters were renamed to immediates.

2) vars vs locals vs registers: suggesting we go with local. Originally I was pushing "register", but that word is probably confusing.

**Comment by nbhuiyan** [comment]
@rwy0717 Ready for another review. There is a merge conflict that I will fix after the changes have been approved.

File: b9/include/b9/instructions.hpp [hunk 7]
```diff
@@ -255,8 +263,8 @@ inline std::ostream &operator<<(std::ostream &out, Instruction i) {
     case OpCode::FUNCTION_CALL:
     case OpCode::PRIMITIVE_CALL:
     case OpCode::JMP:
-    case OpCode::PUSH_FROM_VAR:
-    case OpCode::POP_INTO_VAR:
+    case OpCode::PUSH_FROM_LOCAL:
+    case OpCode::POP_INTO_LOCAL:
```
Comment (rwy7) [inline b9/include/b9/instructions.hpp:267]: Missing push/pop param

Comment (rwy7) [reply b9/include/b9/instructions.hpp:267]: nevermind... how did I miss it :p

Comment (nbhuiyan) [reply b9/include/b9/instructions.hpp:267]: I missed it too at first and only recently pushed that :P

**Review by rwy7 — APPROVED** [review]

**Comment by nbhuiyan** [comment]
@rwy0717 The merge conflicts have been fixed, and the CI build has passed. I think it is ready to be merged.

**Comment by rwy7** [comment]
LGTM. Now that we differentiate between locals and arguments, we should take another look at how the JIT feature flags (passparam, lazy-vm-state) affect how locals and parameters behave. Thanks @nbhuiyan. closes #154.

**Pull request merged**